text/0008-kadalu-rebalance.adoc

```diff
@@ -0,0 +1,132 @@
+
+kadalu volume expand DEV/myvol                  \
+    replica node4.example.com:/exports/myvol/s4 \
+            node5.example.com:/exports/myvol/s5 \
+            node6.example.com:/exports/myvol/s6
+
+----
+
+----
+kadalu volume expand DEV/myvol status
+----
+
+Volume expand API will start a service in the first node (May change this later based on the node online status) from the participating nodes list. Example of the new service:
+
+[source,crystal]
+----
+enum RebalanceCmd
+     None
+     Start
+     Stop
+     Status
+     FixLayoutStart
+     StartForce
+end
+----
+
+----
+glusterfs -s <volfile-server> --volfile-id <volname> --process-name rebalance    \
+	--xlator-option "*distribute.use-readdirp=yes"                           \
+	--xlator-option "*distribute.lookup-unhashed=yes"                        \
+	--xlator-option "*distribute.assert-no-child-down=yes"                   \
+	--xlator-option "replicate.data-self-heal=off"                           \
+	--xlator-option "replicate.metadata-self-heal=off"                       \
+	--xlator-option "replicate.entry-self-heal=off"                          \
+	--xlator-option "*distribute.readdir-optimize=on"                        \
+	--xlator-option "*distribute.rebalance-cmd=<RebalanceCmd.FixLayoutStart> \
+	--xlator-option "*distribute.node-uuid=<node-id>"                        \
+        --xlator-option "*distribute.commit-hash=<unique-id>"                    \
+        -p <pid-file>                                                            \
+        --socket-file <socket-file>                                              \
+        -l <logfile>
+----
+
+== Fix layout Rebalance (Operator based)
+
+Operator parses the CRD modification and understands that Storage Pool expansion is triggered. After starting all the new Server pods. Start a Job that triggers the Fix layout rebalance. kubectl-kadalu can get the status of rebalance by checking the status of the Job.
+
+Refer the above example command to be run as job.
+
+== Files rebalance (Moana based)
+
+This is the manual process can be triggered by admins based on the needs (When no space available in the existing Storage units, or when the Cluster usage is minimal).
+
+"Volume expand" command also prints the note with the steps to run the Rebalance command. Rebalance helps to distribute the existing files uniformly among the Storage units.
+
+----
+kadalu volume rebalance-start <POOL>/<VOLUME>
+----
+
+Stop and Start command will start the rebalance process from the beginning, it doesn't remember the previous progress.
+
+----
+kadalu volume rebalance-stop <POOL>/<VOLUME>
+----
+
+Above command calls the respective API that internally starts a service in respective Storage unit nodes. **Note**: These services will halt once the Rebalance process completes its job. Do not start the service in every nodes of the Volume, start only in the node of the first Storage unit from each distribute groups.
```

Comment:
The design is cool :100: 
Any reason for the command to be run from node of 1st storage unit from each distribute group. Is it avoid redundancy or wrong calculation of starting point of hash range.

Will it not be feasible to call the command from server node as an option, since moana has server-client model and create a node action to fetch necessary xattr details from nodes of 1st storage unit of dist grp to make it much simpler?

Reply:
First node is to make it predictable. Any random node is fine, but Moana should keep a reference where it started the service. We can improve this part later. For now it always starts in the first node of every distribute group.